File: Difficulty: Hard/Edit Distance/edit-distance.java
//{ Driver Code Starts
import java.io.*;
import java.lang.*;
import java.util.*;

class GFG {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int T = Integer.parseInt(br.readLine().trim());
        while (T-- > 0) {
            String s1 = br.readLine();
            String[] S = s1.split(" ");
            String s = S[0];
            String t = S[1];
            Solution ob = new Solution();
            int ans = ob.editDistance(s, t);
            System.out.println(ans);
        }
    }
}

// } Driver Code Ends




class Solution {
    public int editDistance(String word1, String word2) {
        // Code here
        int n=word1.length();
        int m=word2.length();
        int dp[][]=new int[n+1][m+1];

        for(int i=0;i<n+1;i++)
        {
            for(int j=0;j<m+1;j++)
            {
                if(i==0) dp[i][j]=j;
                if(j==0) dp[i][j]=i;
            }
        }
        for(int i=1;i<n+1;i++)
        {
            for(int j=1;j<m+1;j++)
            {
                //same character
                if(word1.charAt(i-1)==word2.charAt(j-1))
                {
                    dp[i][j]=dp[i-1][j-1];
                }
                else{
                    int add=dp[i][j-1];
                    int del=dp[i-1][j];
                    int replace=dp[i-1][j-1];
                    dp[i][j]=Math.min(del,Math.min(add,replace))+1;
                    //dp[i][j]=Math.min(dp[i-1][j],Math.min(dp[i][j-1],dp[i-1][j-1]))+1;
                }
            }
        }
        return dp[n][m];
    }
}

File: Difficulty: Hard/Edit Distance/edit-distance.cpp
//{ Driver Code Starts
#include <bits/stdc++.h>
using namespace std;


// } Driver Code Ends
class Solution {
  public:
    int dp[501][501];
    int solve(int i,int j,string &s,string &t)
    {
        if(i==s.size()){
            return t.size()-j;
        }
        if(j==t.size()){
            return s.size()-i;
        }
        if(dp[i][j]!=-1) return dp[i][j];
        if(s[i]==t[j]){
            return dp[i][j]=solve(i+1,j+1,s,t);
        }
        int replace=1+solve(i+1,j+1,s,t);
        int del=1+solve(i+1,j,s,t);
        int insert=1+solve(i,j+1,s,t);
        return dp[i][j]=min(insert,min(replace,del));
    }
    int editDistance(string str1, string str2) {
        // Code here
        memset(dp,-1,sizeof(dp));
        return solve(0,0,str1,str2);
    }
};

//{ Driver Code Starts.
int main() {
    int T;
    cin >> T;
    while (T--) {
        string s, t;
        cin >> s >> t;
        Solution ob;
        int ans = ob.editDistance(s, t);
        cout << ans << "\n";
    }
    return 0;
}

// } Driver Code Ends